feat: implement low-level dependency injection container

aaronc opened the pull request:
<!--
The default pull request template is for types feat, fix, or refactor.
For other templates, add one of the following parameters to the url:
- template=docs.md
- template=other.md
-->

## Description

closes #9775 
needs #9658

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

blushi commented on container/constructor_info.go:
```diff
@@ -1,22 +1,104 @@
+type ProviderDescriptor struct {
 	// In defines the in parameter types to Fn.
-	In []reflect.Type
+	In []ProviderInput
```
I think I prefer the current naming as it's more consistent with the types names `ProviderInput` and `ProviderOutput`

aaronc replied:
I think I'll change to `Inputs` and `Outputs` for now

blushi commented on container/constructor_info.go:
```diff
@@ -1,22 +1,104 @@
+	Type reflect.Type
+}
+
+func ExtractConstructorInfo(ctr interface{}) (ProviderDescriptor, error) {
```
what about this to be more consistent with the new naming:
```suggestion
func ExtractProviderDescriptor(pvr interface{}) (ProviderDescriptor, error) {
```
I guess we should also update the file name

blushi commented on container/constructor_info.go:
```diff
@@ -1,22 +1,104 @@
+	return expandStructArgsConstructor(rctr)
+}
+
+func doExtractConstructorInfo(ctr interface{}) (ProviderDescriptor, error) {
```
ditto

blushi commented on container/container.go:
```diff
@@ -0,0 +1,336 @@
+)
+
+/*
+TODO:
```
is this still a TODO?

blushi commented on container/simple.go:
```diff
@@ -0,0 +1,60 @@
+)
+
+type simpleProvider struct {
+	ctr    *ProviderDescriptor
```
ditto rename struct field name

blushi commented on container/resolver.go:
```diff
@@ -0,0 +1,11 @@
+package container
+
+import (
+	reflect2 "reflect"
```
why using `reflect2` here?

blushi reviewed (commented):
Just a few nits.

aaronc commented on container/constructor_info.go:
```diff
@@ -1,22 +1,104 @@
+	"reflect"
 
-// ConstructorInfo defines a special constructor type that is defined by
+	"github.com/pkg/errors"
```
It's to wrap the errors with stack traces

blushi reviewed (commented):
I had a more thorough review, focused on understanding the underlying implementation and relationships between the different components, looks good to me, although adding more comments in the code could be useful.

jgimeno commented on container/scope_dep.go:
```diff
@@ -0,0 +1,50 @@
+)
+
+type scopeDepProvider struct {
+	ctr            *ProviderDescriptor
```
for me is more descriptor than provider this var.

jgimeno commented on container/config.go:
```diff
@@ -0,0 +1,178 @@
+	autoGroupTypes   map[reflect.Type]bool
+	onePerScopeTypes map[reflect.Type]bool
+
+	// logging
```
I would extract the logging and the graphing from the options struct. I see the config used only to configure the container. As it is done when constructing with autoGroupTypes and onePerScope types.

aaronc replied:
Where I would put them instead?

aaronc replied:
The alternative would be adding a parameter or overload to `Run` just with debugging options separate from the container options. But that feels relatively involved. If you think it's useful to separate how about we align on an API and do that in a follow-up PR?

aaronc commented:
@blushi @jgimeno I've addressed the review comments. Let me know if you can take a look and we can move forward. @jgimeno is okay to deal with debug options later or leave as is?

aaronc commented:
I've fixed the issues with auto-group and one-per-scope types in #10222. I'm making progress on #9862. Can we merge this now?

I think I might want to pull the logging stuff out of `Option` but let's do that separately.

blushi reviewed (approved):
LGTM

fdymylja reviewed (approved):
LGTM. No strong opinion on https://github.com/cosmos/cosmos-sdk/pull/9666/files#r710071800. But I tend to agree to split logging/graphing functionality from the logic functionality, since overall what happens is not easy to follow I'd try to keep the functionality scope of components as small as possible.
